### PR TITLE
Spike for RequestComparator

### PR DESCRIPTION
**Pull request by C2H6O**
This PR adds a `RequestComparator` that allows for more flexible comparisons of `Request`s and `RequestDescriptor`s. This is especially useful in a MIXED mode when _some_ params/headers might be different (authentication token for example).

### CODE REVIEW

**Comment by speekha** [comment]
First of all, I'm not sure what you're trying to do here? It looks like you're trying to emulate the dynamic mocks with the static ones. The point of static mocks is to be able to provide a fixed scenario based on fixed parameters. If you need some more elaborate rules, you should just use dynamic mocks (plus you can use dynamic and static together: dynamic mocks are evaluated first, then static ones).

Second of all, the code you added doesn't have any tests or documentation for it.

**Comment by C2H6O** [comment]
@speekha yeah, this is definitely not something that I'd want to merge into the repo, was a PR to start a conversation. 

I'm trying to prevent mock data drift from what the backend returns, which means that the mock data should be "refreshed" once in awhile. In order to do that, manual combing through data to figure out that something needs to be moved around in the json files should be minimized.

I'd like to define a set of `headers` and `params` that are important for a particular url so that `Requests` get matched properly.

File: demo/src/main/java/fr/speekha/httpmocker/demo/di/InjectionModule.kt [hunk 1]
```diff
@@ -35,7 +37,11 @@ val injectionModule: Module = module {
     single {
         MockResponseInterceptor.Builder()
             .decodeScenarioPathWith(MirrorPathPolicy())
-            .loadFileWith(get<Context>().assets::open)
+            .loadFileWith(get<Context>().assets::open) { request: Request, requestDescriptor: RequestDescriptor ->
```
[the remark this answers is not in the document]
Comment (C2H6O) [reply demo/src/main/java/fr/speekha/httpmocker/demo/di/InjectionModule.kt]: Yeah, you are right. I wanted to illustrate how this method _would_ work in general. The logic mirrors what you had in the default comparator.

**Comment by speekha** [comment]
Could you elaborate a bit on your issue? It seems to me you're saying some query parameters or headers should be ignored because they might change on the backend and you don't want that to mess up with you scenarios. Did I understand that right? If that's the case, if they're not important for your match, you can just remove them from your scenarios and they won't be checked when matching. 
Also, following the issue you opened previously, I've added the possibility to force exact matches (all parameters and headers have to match the ones declared in the json file, missing ones in the scenario are treated like a failure and not ignored). So in your previous case, if you'd add "exact-match" : true in the request without a token, that request would only match requests that don't have the extra param.

**Comment by speekha** [comment]
@C2H6O I've extracted the comparison in a separate class (not configurable yet, though, I'm still not entirely convinced that's useful). I've also added a new element: if you set some parameter or header to `null` in your scenario, then that parameter or header will have to be absent from the request for it to match.

**Comment by speekha** [comment]
@C2H6O I published a new version with the changes I told you about : 
* possibility of exact match (no extra param or header allowed compared to what's in the scenario)
* match absent parameters (params or headers explicitly set to false must not be present in a request to match)
Let me know if that answers your needs. If not, let's discuss what improvements we could come up with to address those.
Without a reply from you in the next few days, I'll close this PR.

**Pull request closed without merging**